### PR TITLE
Use a free function for `_log_from_c`

### PR DESCRIPTION
**Pull request by eric-wieser**
As raised in gh-1212, there is no need to use subclassing here, and doing so removes flexibility from the user.

After this change, `libgpi` logging will work independently of whatever logging class is configured via the builtin `logging.setLoggerClass`. More importantly, it removes one more reason for `SimBaseLog` to exist - the ultimate goal is to remove that class entirely, since it's behavior is hostile to users wanting to use the logging module normally.

This also fixes a bug where the name argument of `gpilog` is ignored.

### CODE REVIEW

**Comment by eric-wieser** [comment]
I'll improve the commit message when I merge, if that works for you

**Comment by eric-wieser** [comment]
Sorry @imphil, ended up making some more substantial changes after having thought about this some more. Wasn't really a fan of putting these functions in `__init__`

**Comment by ktbarrett** [comment]
I preferred the previous implementation. When we add #1225, we will need the whole interface required by C to be defined in the entry module.

**Comment by eric-wieser** [comment]
@ktbarrett: perhaps I'm missing something, but I don't think there is a good reason to allow the user to configure gpilog to do anything other than hook up to the python logging infrastructure, so that needn't be party of a custom entry point.

**Comment by eric-wieser** [comment]
I could put them in `cocotb._gpilog` if that would make more sense to you.

**Comment by ktbarrett** [comment]
When I say custom entry point I mean a modular interface. You could wipe away the cocotb python module and put something totally different in place, and it will work. So saying you have a custom entry point, but will always use `cocotb.log` for the gpilog is floating in a weird halfway between a proper modular layered design and the monolith we have now.

And that's why I want those functions at known location in the entry module. Not doing so prevents people from taking advantage of a (future) modular design and offers us nothing.

**Comment by eric-wieser** [comment]
I assume that all you object to in the second revision of this patch is the file which defines these functions.

I would argue then that until we have a clear design for this modularity (perhaps a wiki page), we should stick to keeping all logging-related python code in a single file. If the functions are private, as they are here, we can move them later without concern for compatibility.

Making these functions easy for a future maintainer to find is more important than optimizing for a design we don't have yet.

**Comment by eric-wieser** [comment]
Or perhaps a simpler argument: these functions have not moved file - moving file is out of scope for this patch, let's move them in a subsequent one with suitable motivation in the commit message.

**Comment by ktbarrett** [comment]
I have no problem it being tomorrow's problem. Just seeing if it was worth the effort today.

File: cocotb/share/lib/embed/gpi_embed.c [hunk 1]
```diff
@@ -259,52 +259,47 @@ int embed_sim_init(gpi_sim_info_t *info)
...
 
     cocotb_module = NULL;
-    simlog_obj = NULL;
 
     // Ensure that the current thread is ready to call the Python C API
     PyGILState_STATE gstate = PyGILState_Ensure();
     to_python();
 
-    if (get_module_ref(COCOTB_MODULE, &cocotb_module))
+    if (get_module_ref("cocotb", &cocotb_module))
```
Comment (imphil) [inline cocotb/share/lib/embed/gpi_embed.c]: What's the motivation for this change? For consistency with `cocotb.log` below? If so, please remove the definition of COCOTB_MODULE.

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 2]
```diff
@@ -207,7 +212,7 @@ void gpi_log(const char *name, long level, const char *pathname, const char *fun
     }
 
     // Log function args are level, filename, lineno, msg, function
```
Comment (imphil) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: Please adjust this comment for the added name argument.

File: cocotb/share/lib/embed/gpi_embed.c (hunk 1 above)
Comment (eric-wieser) [reply cocotb/share/lib/embed/gpi_embed.c]: Seemed like a fairly useless constant given its defined as a shorter string than its name. I'll come back and remove the env bar.

File: cocotb/share/lib/gpi_log/gpi_logging.c (hunk 2 above)
Comment (eric-wieser) [reply cocotb/share/lib/gpi_log/gpi_logging.c]: Will do

**Comment by eric-wieser** [comment]
Comments addressed, will merge when CI passes
